File: 3-Greedy Algorithms, Minimum Spanning Tree, and Dynamic Programming/1-Greedy Algorithm/3-Minimum Spanning Tree Problem/Python/union_find.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
A simple implementation of UnionFind data structure.
Maintain a partition of a set of objects
"""

__author__ = 'Ziang Lu'


class IllegalArgumentError(ValueError):
    pass


class UnionFindObj(object):
    def __init__(self):
        self._leader = self

    @property
    def name(self):
        """
        Accessor of name.
        :return: str
        """
        pass

    @property
    def leader(self):
        """
        Accessor of leader.
        :return: UnionFindObj
        """
        return self._leader

    @leader.setter
    def leader(self, leader):
        """
        Mutator of leader.
        :param leader: UnionFindObj
        :return: None
        """
        self._leader = leader


class UnionFind(object):
    def __init__(self, objs):
        """
        Constructor with parameter.
        :param objs: list[UnionFindObj]
        """
        self._groups = {}
        for obj in objs:
            self._groups[obj.name] = [obj]

    def find(self, obj):
        """
        Returns the name of the group that the given object belongs to.
        :param obj: UnionFindObj
        :return: str
        """
        return obj.leader.name
        # Running time complexity: O(1)

    def union(self, group_name_a, group_name_b):
        # Check whether the input strings are null or empty
        if group_name_a is None or len(group_name_a) == 0 or \
                group_name_b is None or len(group_name_b) == 0:
            raise IllegalArgumentError('The input strings should not be None or'
                                       ' empty.')
        # Check whether the input group names exist
        if group_name_a not in self._groups or group_name_b not in self._groups:
            raise IllegalArgumentError("The input group names don't both "
                                       "exist.")

        group_a, group_b = \
            self._groups[group_name_a], self._groups[group_name_b]
        # In order to reduce the number of leader updates, let the smaller group
        # inherit the leader of the larger one.
        if len(group_a) >= len(group_b):
            larger, smaller = group_a, group_b
        else:
            larger, smaller = group_b, group_a
        larger_leader = larger[0].leader
        smaller_name = smaller[0].leader.name
        self._update_leader(group=smaller, new_leader=larger_leader)
        larger.extend(smaller)
        self._groups.pop(smaller_name)

    def _update_leader(self, group, new_leader):
        """
        Private helper function to update the leader of the given group to the
        given new leader.
        :param group: list[UnionFindObj]
        :param new_leader: UnionFindObj
        :return: None
        """
        for obj in group:
            obj.leader = new_leader
        # Running time complexity: O(1)

File: 3-Greedy Algorithms, Minimum Spanning Tree, and Dynamic Programming/1-Greedy Algorithm/3-Minimum Spanning Tree Problem/Java/union_find/UnionFind.java
package union_find;

import java.util.ArrayList;
import java.util.HashMap;

/**
 * Simple implementation of Union-Find data structure.
 * Maintain a partition of a set of objects
 *
 * @param T object type that implements UnionFindObj interface
 * @author Ziang Lu
 */
public class UnionFind <T> {

    /**
     * Mapping between group names and the corresponding group.
     * Maintain a linked structure, and each subset has an arbitrary leader
     * (representative of the group) object, and the group name is exactly the
     * name of the leader.
     */
    private HashMap<String, ArrayList<UnionFindObj>> groups;

    /**
     * Constructor with parameter.
     * @param objs list of sole objects
     */
    public UnionFind(ArrayList<T> objs) {
        groups = new HashMap<String, ArrayList<UnionFindObj>>();
        for (T obj : objs) {
            UnionFindObj o = (UnionFindObj) obj;
            ArrayList<UnionFindObj> group = new ArrayList<UnionFindObj>();
            group.add(o);
            groups.put(o.objName(), group);
        }
    }

    /**
     * Returns the number of groups.
     * @return number of groups
     */
    public int numOfGroups() {
        return groups.size();
    }

    /**
     * Returns the name of the group, which is exactly the name of the group
     * leader, that the given object belongs to.
     * @param obj given object
     * @return name of the group, which is exactly the name of the group leader
     */
    public String find(UnionFindObj obj) {
        return obj.leader().objName();
        // Running time complexity: O(1)
    }

    /**
     * Fuses the given two groups together.
     * Objects in the first group and objects in the second group should all
     * coalesce, and be now in one single group.
     * @param groupNameA name of the first group
     * @param groupNameB name of the second group
     */
    public void union(String groupNameA, String groupNameB) {
        // Check whether the input strings are null or empty
        if ((groupNameA == null) || (groupNameA.length() == 0) || (groupNameB == null) || (groupNameB.length() == 0)) {
            throw new IllegalArgumentException("The input group names should not be null or empty.");
        }
        // Check whether the input group names exist
        if (!(groups.containsKey(groupNameA)) || !(groups.containsKey(groupNameB))) {
            throw new IllegalArgumentException("The input group names don't both exist.");
        }

        ArrayList<UnionFindObj> groupA = groups.get(groupNameA), groupB = groups.get(groupNameB);
        // In order to reduce the number of leader updates, let the smaller group inherit the leader of the larger one.
        ArrayList<UnionFindObj> larger = null, smaller = null;
        if (groupA.size() >= groupB.size()) {
            larger = groupA; smaller = groupB;
        } else {
            larger = groupB; smaller = groupA;
        }
        UnionFindObj largerLeader = larger.get(0).leader();
        String smallerName = smaller.get(0).objName();
        updateLeader(smaller, largerLeader);

        larger.addAll(smaller);
        groups.remove(smallerName);
        // Running time complexity: O(n)
    }

    /**
     * Private helper method to update leader of the given group to the given new leader.
     * @param group given group
     * @param newLeader given new leader
     */
    private void updateLeader(ArrayList<UnionFindObj> group, UnionFindObj newLeader) {
        for (UnionFindObj obj : group) {
            obj.setLeader(newLeader);
        }
        // Running time complexity: O(n)
    }

}
